src/bp/core/services/dialog/flow/service.ts

```diff
@@ -155,36 +172,68 @@ export class ScopedFlowService {
     private logger: Logger,
     private moduleLoader: ModuleLoader,
     private realtime: RealtimeService,
-    private botService: BotService
-  ) {}
+    private botService: BotService,
+    private invalidateFlow: (key: string, flow?: FlowView, newKey?: string) => void
+  ) {
+    this.cache = new ArrayCache<string, FlowView>(
+      x => x.name,
+      // TODO not sure about this
+      (x, pkey, nkey) => ({ ...x, name: nkey, location: nkey })
+    )
+    this.expectedSavesCache = new LRUCache({ max: 100, maxAge: ms('20s') })
+  }
 
-  async handleInvalidatedCache(flowName: string) {
-    if (this._flowCache.isEmpty()) {
+  public localInvalidateFlow(key: string, flow?: FlowView, newKey?: string) {
+    if (!this.cache.values().length) {
       return
     }
 
-    const flowPath = this.toFlowPath(flowName)
+    if (flow) {
+      this.cache.update(key, flow)
+    } else if (newKey) {
+      this.cache.rename(key, newKey)
+    } else {
+      this.cache.remove(key)
+    }
+  }
 
-    if (await this.ghost.fileExists(FLOW_DIR, flowPath)) {
-      const flow = await this.parseFlow(flowPath)
+  public async handleInvalidatedCache(flowName: string) {
+    const flowPath = this.toFlowPath(flowName)
+    const expectedSaves = this.expectedSavesCache.get(flowPath)
 
-      this._flowCache.upsertFlow(this.botId, flow)
+    if (!expectedSaves) {
+      if (await this.ghost.fileExists(FLOW_DIR, flowPath)) {
+        const flow = await this.parseFlow(flowPath)
+        this.invalidateFlow(flowPath, flow)
+      } else {
+        this.invalidateFlow(flowPath, undefined)
+      }
     } else {
-      this._flowCache.deleteFlow(this.botId, flowPath)
+      this.expectedSavesCache.set(flowPath, expectedSaves - 1)
     }
 
+    /* TODO : is this still necessary?
+
     // parent flows are only used by the NDU
     if (this._isOneFlow()) {
       const flows = this._flowCache.get(this.botId)
       const flowsWithParents = this.addParentsToFlows(flows)
 
       this._flowCache.set(this.botId, flowsWithParents)
     }
+
+    */
+  }
+
+  private incrementExpectedSaves(flowName: string) {
+    const current = this.expectedSavesCache.get(flowName)
+    // we increment by 2 because we always write the .flow and .ui at the same time
+    this.expectedSavesCache.set(flowName, current === undefined ? 2 : current + 2)
   }
```

Comment:
This might need a bit of explaining. Basically whenever the flow services updates a files, it notes that it's going to receive at least 2 file update events (one for .ui, one for .flow). The next 2 file update events can then be safely ignored because they would be redundant.